scripts/simulateShortcut.ts

```diff
@@ -117,31 +175,93 @@ async function simulateShortcutOnQuoter(
     const index = quoteTokens.findIndex((q) => q === t);
     report.dust[t] = quote.amountOut[index];
   });
-  console.log('Simulation: ', report);
+  if (!isRecursiveCall) {
+    console.log('Simulation: ', report);
+  }
+
+  return report;
 }
 
-async function simulateOnForge(
+async function simulateShortcutOnForge(
   shortcut: Shortcut,
   chainId: ChainIds,
   amountsIn: string[],
+  script: WeirollScript,
+  tokensIn: AddressArg[],
+  tokensOut: AddressArg[],
+  slippage: BigNumber,
   forgePath: string,
   rpcUrl: string,
   blockNumber: number,
   roles: SimulationRoles,
-): Promise<void> {
-  const { script, metadata } = await shortcut.build(chainId);
-
-  const { tokensIn, tokensOut } = metadata;
-  if (!tokensIn || !tokensOut) throw 'Error: Invalid builder metadata';
-  if (amountsIn.length != tokensIn.length)
-    throw `Error: Incorrect number of amounts for shortcut. Expected ${tokensIn.length}`;
-
+  isRecursiveCall = false,
+): Promise<Report> {
   const { commands, state } = script;
 
-  const provider = new StaticJsonRpcProvider(rpcUrl);
-  const weirollWallet = await getNextWeirollWalletFromRecipeMarketHub(provider, roles.recipeMarketHub.address!);
-  roles.weirollWallet = { address: weirollWallet, label: 'WeirollWallet' };
-  const txData = await generateMulticallTxData(shortcut, chainId, commands, state, roles.recipeMarketHub.address!);
+  const shortcutExecutionMode = getShortcutExecutionMode(shortcut, chainId);
+
+  let txData: string;
+  let forgeContract: string;
+  let forgeContractABI: Record<string, unknown>[];
+  let forgeTest: string;
+  let forgeTestRelativePath: string;
+  switch (shortcutExecutionMode) {
+    case ShortcutExecutionMode.MULTICALL__AGGREGATE: {
+      const provider = new StaticJsonRpcProvider(rpcUrl);
+      let minAmountOut = DEFAULT_SETTER_MIN_AMOUNT_OUT;
+      // NB: simulate first with `minAmountOut` set to '1' wei and get the actual `amountOut` from quoter.
+      // Then, calculate the expected `minAmountOut` after applying maximum slippage, and finally simulate again.
+      if (!isRecursiveCall) {
```

Comment:
LGTM, just unsure if it's useful to have this logic into these simulate scripts. Since is a simulation, setting it to 1 it's fine, and can be used just to get the minAmountOut and have a separate function to generate this calldata

https://linear.app/enso/issue/INT-405/generate-multicall-calldata-with-minamountout-set

Anyhow the ticket above needs to be added, up to you if we want to have a "simple" simulation and a separate simulation for the multicall flow.

Reply:
By default setter shortcuts are run with `minAmountOut=1`.
If shortcuts are run with `--slippage=0`, `minAmountOut=1`.
If shortcuts are run with `--slippage=<[0,10000]>`, then the shortcut runs twice. It first gets the amount out, calculates the new `minAmountOut` applying the desired slippage (in BIPS), and finally runs the shortcut again with the new `minAmountOut`.